File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/sequence/AnimationSequence.java
package at.wirecube.additiveanimations.additive_animator.sequence;

import java.util.Arrays;
import java.util.List;

public abstract class AnimationSequence {

    public abstract void start();
    public abstract void setDelayInSequence(long delay);
    public abstract long getTotalDurationInSequence();

    public static AnimationSequence playTogether(AnimationSequence... animations) {
        return new PlayTogetherAnimationSequence(Arrays.asList(animations));
    }

    public static AnimationSequence playTogether(List<AnimationSequence> animations) {
        return new PlayTogetherAnimationSequence(animations);
    }

    public static AnimationSequence playSequentially(AnimationSequence... animations) {
        return new PlaySequentiallyAnimationSequence(Arrays.asList(animations));
    }

    public static AnimationSequence playSequentially(List<AnimationSequence> animations) {
        return new PlaySequentiallyAnimationSequence(animations);
    }

    public static AnimationSequence playWithDelayBetweenAnimations(long stagger, AnimationSequence... animations) {
        return new PlayWithStaggerAnimationSequence(stagger, animations);
    }
}

File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/sequence/AnimationSequence.kt
package at.wirecube.additiveanimations.additive_animator.sequence

abstract class AnimationSequence {

    abstract fun start()
    abstract fun setDelayInSequence(delay: Long)
    abstract fun getTotalDurationInSequence(): Long

    companion object {
        @JvmStatic
        fun playTogether(vararg animations: AnimationSequence): AnimationSequence {
            return PlayTogetherAnimationSequence(animations.toList())
        }

        @JvmStatic
        fun playTogether(animations: List<AnimationSequence>): AnimationSequence {
            return PlayTogetherAnimationSequence(animations)
        }

        @JvmStatic
        fun playSequentially(vararg animations: AnimationSequence): AnimationSequence {
            return PlaySequentiallyAnimationSequence(animations.toList())
        }

        @JvmStatic
        fun playSequentially(animations: List<AnimationSequence>): AnimationSequence {
            return PlaySequentiallyAnimationSequence(animations)
        }

        @JvmStatic
        fun playWithDelayBetweenAnimations(stagger: Long, vararg animations: AnimationSequence): AnimationSequence {
            return PlayWithStaggerAnimationSequence(stagger, *animations)
        }
    }
}



